curp/src/bg_tasks.rs

```diff
@@ -0,0 +1,635 @@
+    let resp = connect.append_entries(req, RPC_TIMEOUT).await;
+
+    #[allow(clippy::unwrap_used)]
+    // indexing of `next_index` or `match_index` won't panic because we created an entry when initializing the server state
+    match resp {
+        Err(e) => warn!("append_entries error: {}", e),
+        Ok(resp) => {
+            let resp = resp.into_inner();
+            // calibrate term
+            let state = state.upgradable_read();
+            if resp.term > state.term {
+                let mut state = RwLockUpgradableReadGuard::upgrade(state);
+                state.update_to_term(resp.term);
+                return;
+            }
+            if !resp.success {
+                let mut state = RwLockUpgradableReadGuard::upgrade(state);
+                *state.next_index.get_mut(&connect.addr).unwrap() -= 1;
+            }
+        }
+    };
+}
+
+/// Background apply
+async fn bg_apply<C: Command + 'static, CE: 'static + CommandExecutor<C>>(
+    state: Arc<RwLock<State<C>>>,
+    cmd_executor: Arc<CE>,
+    comp_chan: SpmcKeyBasedSender<C::K, SyncCompleteMessage<C>>,
+    spec: Arc<Mutex<SpeculativePool<C>>>,
+) {
+    let commit_trigger = state.read().commit_trigger();
+    loop {
+        // wait until there is something to commit
+        let state = loop {
+            {
+                let state = state.upgradable_read();
+                if state.need_commit() {
+                    break state;
+                }
+            }
+            commit_trigger.listen().await;
+        };
+
+        let mut state = RwLockUpgradableReadGuard::upgrade(state);
+        #[allow(clippy::integer_arithmetic, clippy::indexing_slicing)]
+        // TODO: overflow of log index should be prevented
+        for i in (state.last_applied + 1)..=state.commit_index {
+            for cmd in state.log[i].cmds().iter() {
+                // TODO: execution of leaders and followers should be merged
+                // leader commits a log by sending it through compl_chan
+                if state.is_leader() {
+                    if comp_chan
+                        .send(
+                            cmd.keys(),
+                            SyncCompleteMessage::new(i.numeric_cast(), Arc::clone(cmd)),
+                        )
+                        .is_err()
+                    {
+                        error!("The comp_chan is closed on the remote side");
+                        break;
+                    }
+                } else {
+                    // followers execute commands directly
+                    let cmd_executor = Arc::clone(&cmd_executor);
+                    let cmd = Arc::clone(cmd);
+                    let spec = Arc::clone(&spec);
+                    let _handle = tokio::spawn(async move {
+                        // TODO: execute command in parallel
+                        // TODO: handle execution error
+                        let _execute_result = cmd.execute(cmd_executor.as_ref()).await;
```

Comment:
Can we make sure the first-come command executed first? What if there are conflicted commands?

Reply:
Fixed in new commit.